<a href="https://colab.research.google.com/github/wendirad/week1/blob/main/Task_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import pandas as pd
import numpy as np
import pandas_datareader.data as web
import matplotlib.pyplot as plt
from textblob import TextBlob

In [ ]:
AAPL = pd.read_csv("/content/AAPL_historical_data.csv")
news = pd.read_csv('/content/raw_analyst_ratings.csv')

In [ ]:
AAPL.head()

Date      Open      High       Low     Close  Adj Close     Volume  \
0  1980-12-12  0.128348  0.128906  0.128348  0.128348   0.098943  469033600   
1  1980-12-15  0.122210  0.122210  0.121652  0.121652   0.093781  175884800   
2  1980-12-16  0.113281  0.113281  0.112723  0.112723   0.086898  105728000   
3  1980-12-17  0.115513  0.116071  0.115513  0.115513   0.089049   86441600   
4  1980-12-18  0.118862  0.119420  0.118862  0.118862   0.091630   73449600   

   Dividends  Stock Splits  
0        0.0           0.0  
1        0.0           0.0  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0

In [ ]:
news.head()

Unnamed: 0                                           headline  \
0           0            Stocks That Hit 52-Week Highs On Friday   
1           1         Stocks That Hit 52-Week Highs On Wednesday   
2           2                      71 Biggest Movers From Friday   
3           3       46 Stocks Moving In Friday's Mid-Day Session   
4           4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

                        date stock  
0  2020-06-05 10:30:54-04:00     A  
1  2020-06-03 10:45:20-04:00     A  
2  2020-05-26 04:30:07-04:00     A  
3  2020-05-22 12:45:06-04:00     A  
4  2020-05-22 11:38:59-04:00     A

In [ ]:
len(news)

1407328

In [ ]:
len(AAPL)

10998

To merge or align these two datasets (stock data and news data), need to make sure that the **date** column in both datasets is consistent in format. In your case:

1. **Stock Data Date Format**: The date format is `YYYY-MM-DD` (e.g., `1980-12-12`).
2. **News Data Date Format**: The date format is `YYYY-MM-DD HH:MM:SS-04:00` (e.g., `2020-06-05 10:30:54-04:00`), which includes a timestamp and timezone information.


To normalize dates and align the news dataset with the stock trading dataset, need to have both datasets dates that correspond to the same format, and the time component in the news data should be either ignored or adjusted

In [ ]:
# Need to make column 'date' of both datas have the same format
news['date'] = pd.to_datetime(news['date'].astype(str), errors='coerce').dt.tz_localize(None).dt.date
AAPL['Date'] = pd.to_datetime(AAPL['Date'], errors='coerce').dt.date
news[['date']].head(), AAPL[['Date']].head()

(         date
 0  1980-12-12
 1  1980-12-15
 2  1980-12-16
 3  1980-12-17
 4  1980-12-18,
          Date
 0  1980-12-12
 1  1980-12-15
 2  1980-12-16
 3  1980-12-17
 4  1980-12-18)

In [ ]:
merged_data = pd.merge(AAPL, news, left_on='Date', right_on='date', how='inner')
merged_data.head()

Date      Open      High       Low     Close  Adj Close     Volume  \
0  1980-12-12  0.128348  0.128906  0.128348  0.128348   0.098943  469033600   
1  1980-12-15  0.122210  0.122210  0.121652  0.121652   0.093781  175884800   
2  1980-12-16  0.113281  0.113281  0.112723  0.112723   0.086898  105728000   
3  1980-12-17  0.115513  0.116071  0.115513  0.115513   0.089049   86441600   
4  1980-12-18  0.118862  0.119420  0.118862  0.118862   0.091630   73449600   

   Dividends  Stock Splits  Unnamed: 0  \
0        0.0           0.0           0   
1        0.0           0.0           1   
2        0.0           0.0           2   
3        0.0           0.0           3   
4        0.0           0.0           4   

                                            headline  \
0            Stocks That Hit 52-Week Highs On Friday   
1         Stocks That Hit 52-Week Highs On Wednesday   
2                      71 Biggest Movers From Friday   
3       46 Stocks Moving In Friday's Mid-Day Session   
4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

         date stock  
0  1980-12-12     A  
1  1980-12-15     A  
2  1980-12-16     A  
3  1980-12-17     A  
4  1980-12-18     A

In [ ]:
merged_data.drop(columns=['Unnamed: 0','Dividends','Stock Splits'],inplace=True)

merged_data.head()

Date      Open      High       Low     Close  Adj Close     Volume  \
0  1980-12-12  0.128348  0.128906  0.128348  0.128348   0.098943  469033600   
1  1980-12-15  0.122210  0.122210  0.121652  0.121652   0.093781  175884800   
2  1980-12-16  0.113281  0.113281  0.112723  0.112723   0.086898  105728000   
3  1980-12-17  0.115513  0.116071  0.115513  0.115513   0.089049   86441600   
4  1980-12-18  0.118862  0.119420  0.118862  0.118862   0.091630   73449600   

                                            headline  \
0            Stocks That Hit 52-Week Highs On Friday   
1         Stocks That Hit 52-Week Highs On Wednesday   
2                      71 Biggest Movers From Friday   
3       46 Stocks Moving In Friday's Mid-Day Session   
4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

         date stock  
0  1980-12-12     A  
1  1980-12-15     A  
2  1980-12-16     A  
3  1980-12-17     A  
4  1980-12-18     A

## **Sentiment Analysis**
#### **TextBlob**
 Widely used for basic sentiment analysis tasks and can classify sentiment into categories like positive, negative, or neutral, or provide a sentiment score indicating the overall tone.

In [ ]:
news['headline'] = pd.DataFrame(news['headline'])

In [ ]:
def get_sentiment(text):
    blob = TextBlob(text)
    return blob.sentiment.polarity

# Apply sentiment analysis to the headlines
merged_data['sentiment_score'] = news['headline'].apply(get_sentiment)


In [ ]:
merged_data['sentiment_score'].tail()

10993    0.068182
10994    0.068182
10995    0.000000
10996    0.000000
10997    0.500000
Name: sentiment_score, dtype: float64

The **sentiment scores** you've provided indicate the **polarity** or tone of the news headlines for those specific records, where a score closer to +1 signifies a more **positive** sentiment, a score of 0 represents a **neutral** sentiment, and a score closer to -1 indicates a **negative** sentiment.

Here's a breakdown of your data:

```
Index    Sentiment Score
10993     0.068182  (Neutral to slightly positive)
10994     0.068182  (Neutral to slightly positive)
10995     0.000000  (Neutral)
10996     0.000000  (Neutral)
10997     0.500000  (Moderately positive)
```

### Key Observations:
1. **Index 10993 and 10994**: The sentiment score of **0.068** suggests that the headlines associated with these indexes are **slightly positive**, but the sentiment is very close to neutral.
2. **Index 10995 and 10996**: These scores of **0.000000** indicate a **neutral sentiment**, meaning the headlines are not leaning towards either positive or negative emotions.
3. **Index 10997**: The sentiment score of **0.5** suggests a **moderately positive sentiment**, meaning the headline contains more positive tone or content.

### What can we say?
- Overall, the sentiment across these headlines is generally **neutral to slightly positive** with the exception of **Index 10997**, which shows a moderately positive sentiment.
- The sentiment distribution suggests there isn't any extreme polarity (like negative or very positive), indicating that these headlines might mostly involve neutral reports with a few that contain positive undertones.

## **Analysis**
Calculating the **Daily Stock Returns**, which represent the percentage change in the stock's closing price from one day to the next,.



In [ ]:
merged_data['Daily_Return'] =merged_data['Close'].pct_change() * 100

# Display the results
print(merged_data[['Date', 'Close', 'Daily_Return']].head())

         Date     Close  Daily_Return
0  1980-12-12  0.128348           NaN
1  1980-12-15  0.121652     -5.217061
2  1980-12-16  0.112723     -7.339788
3  1980-12-17  0.115513      2.475091
4  1980-12-18  0.118862      2.899246


Looking at the **Daily_Return** values, here's a summary of what can be inferred:

1. ** Day (1980-12-15)**:  
   - The **Daily_Return** is `-5.22%`.  
   - The stock price dropped significantly from `0.128348` to `0.121652`, representing a decrease of about 5.22%.


### Key Observations:
- There were two consecutive days of significant negative returns (`-5.22%` and `-7.34%`), indicating a sharp decline in stock price.
- Following the drop, there was a recovery in stock price with two days of positive returns (`+2.48%` and `+2.90%`), though the positive returns are smaller compared to the initial drop.
- The stock experienced high volatility over the short period, moving from a sharp decline to a mild recovery.

## **Correlation Analysis**
Pearson correlation coefficient between average daily sentiment scores and stock daily returns.
### Pearson Correlation

The **Pearson correlation coefficient** is a measure of the linear relationship between two variables. It ranges from **-1 to 1**, where:
- **+1** indicates a **perfect positive linear relationship** (as one variable increases, the other also increases in a perfectly linear fashion).
- **-1** indicates a **perfect negative linear relationship** (as one variable increases, the other decreases in a perfectly linear fashion).
- **0** indicates **no linear relationship**.

In [ ]:
correlation = merged_data[['sentiment_score', 'Daily_Return']].corr(method='pearson')
correlation

sentiment_score  Daily_Return
sentiment_score         1.000000      0.005211
Daily_Return            0.005211      1.000000

## **Result**
The Pearson correlation coefficient between sentiment_score and Daily_Return is 0.005211, which indicates a very weak positive correlation between the two variables. The sentiment score doesn't have a strong or meaningful correlation with daily stock returns in this dataset.